ENH: stats: Crystalball distribution

thomaskeck opened the pull request:
The crystalball function is used in high-energy physics to parametrize data.
It is basically a Gaussian with a one-sided power-law tail.
https://en.wikipedia.org/wiki/Crystal_Ball_function

This is a split of, of another PR which contained several other distributions https://github.com/scipy/scipy/pull/6795

ev-br commented:
One issue with tests which fail is that higher moments do not exist for n=3.
It'd be best to calculate the expressions for the moments analytically, and add a `_munp` method to return either a correct value or an inf.

thomaskeck commented:
I'm working on the _munp implementation,
but it is more complex than I expected.

Calculating the formulas using mathematica was easy, but there are numerical instabilities and odd definitions of the functions which are required to calculate the n-th moment (e.g. hyp2f1).

So if munp is required I will need some more time (maybe I find some time next weekend to dive deeper into the mathematics behind hyp2f1).

ev-br commented:
From a 10000 feet: Mathematica likes to express things in terms of hyp2f1, even if there's a simpler expression available (or at least it did in version 3 or something). Moments can be left for later I think.

thomaskeck commented:
Fixed merge conflicts.
Implemented _munp.

@ev-br 
You were correct the hyp2f1 function were unnecessary, I calculated the integral by hand
using some substitutions and the binomial expansion; the unittests check the calculated
values from _munp against the values calculated by mathematica (wolframalpha.com).

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -5206,6 +5206,92 @@ def _entropy(self, beta):
+
+    %(example)s
+    """
+    def _pdf(self, x, alpha, n):
```
Maybe better use some other variable name, https://github.com/scipy/scipy/issues/5982

thomaskeck replied:
I renamed alpha to beta and n to m.

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -5206,6 +5206,92 @@ def _entropy(self, beta):
+        N = 1.0 / (n/alpha / (n-1) * np.exp(-alpha**2 / 2.0) + _norm_pdf_C * _norm_cdf(alpha))
+        return N * _lazywhere(np.atleast_1d(x > -alpha), (x, alpha, n),
+                              f = lambda x, alpha, n: np.exp(-x**2 / 2),
+                              f2 = lambda x, alpha, n: (n/alpha)**n * np.exp(-alpha**2 / 2.0) * (n/alpha - alpha - x)**(-n))
```
Line way too long, here and elsewhere.

thomaskeck replied:
I refactored the lazywhere and introduced a right-hand-side (rhs) and left-hand-side (lhs)
lambda function which is passed to f and f2 of _lazywhere.
The lines are now shorter.

thomaskeck commented:
Ping.

I fixed the latest merge conflicts, are there further suggestions or comments to this PR?

thomaskeck commented:
Pong?

thomaskeck commented:
Hello, this is my bi-monthly reminder for this PR.
I understand that the PR is probably forgotten and not shown anymore on the first page of open PRs.
But still I try to get some attention by this comment :-)
@ev-br Since you are provided me feedback in the past I mention you directly, which I should have done maybe earlier :-)

pv commented:
Looks good to me now. Thanks, merging!

pv commented:
Some issues remaining: gh-7687
